core-kotlin-modules/core-kotlin-datastructures/src/test/kotlin/com/baeldung/frequencymap/FrequencyMapUnitTest.kt

```diff
@@ -0,0 +1,123 @@
+        assertEquals(expectedMap, frequencyMapUsingMutableMap(list))
+    }
+
+    @Test
+    fun `test frequency map using groupingBy() method`() {
+        val list = listOf(1, 2, 1, 3, 2, 4, 4, 7,9,7,3,2,1)
+        val expectedMap = mapOf(1 to 3, 2 to 3, 3 to 2, 4 to 2, 7 to 2, 9 to 1)
+        val actualMap = list.groupingBy { it }.eachCount()
+        assertEquals(expectedMap, actualMap)
+    }
+
+    @Test
+    fun `test frequency map using frequency() method`() {
+        val list = listOf(1, 2, 1, 3, 2, 4, 4, 7,9,7,3,2,1)
+        val expectedMap = mapOf(1 to 3, 2 to 3, 3 to 2, 4 to 2, 7 to 2, 9 to 1)
+        val actualMap = mutableMapOf<Int, Int>()
+        for(value in list.distinct()){
+            actualMap[value] = Collections.frequency(list, value)
+        }
+        assertEquals(expectedMap, actualMap)
+    }
+
+    @Test
+    fun `test frequency map using treemap method`() {
+        val list = listOf(1, 2, 1, 3, 2, 4, 4, 7,9,7,3,2,1)
+        val expectedMap = mapOf(1 to 3, 2 to 3, 3 to 2, 4 to 2, 7 to 2, 9 to 1)
+        val actualMap = frequencyMapUsingTreeMapMethod(list)
+        val sortedList = list.sorted().distinct()
+
+        assertEquals(expectedMap, actualMap)
+        assertEquals(sortedList, actualMap.keys.toList())
+    }
+
+    @Test
+    fun `test frequency map using binary search method`() {
+        val list = listOf(1, 2, 1, 3, 2, 4, 4, 7,9,7,3,2,1)
+        val expectedMap = mapOf(1 to 3, 2 to 3, 3 to 2, 4 to 2, 7 to 2, 9 to 1)
+        val actualMap = frequencyMapUsingBinarySearchMethod(list)
+
+        assertEquals(expectedMap, actualMap)
+    }
+
+    @Test
+    fun `test frequency map using hashset method`() {
+        val list = listOf(1, 2, 1, 3, 2, 4, 4, 7,9,7,3,2,1)
+        val expectedMap = mapOf(1 to 3, 2 to 3, 3 to 2, 4 to 2, 7 to 2, 9 to 1)
+        val actualMap = FrequencyMapUsingHashSetMethod(list)
+
+        assertEquals(expectedMap, actualMap)
+    }
+
+    @Test
+    fun `test frequency map using merge() method`() {
+        val list = listOf(1, 2, 1, 3, 2, 4, 4, 7,9,7,3,2,1)
+        val expectedMap = mapOf(1 to 3, 2 to 3, 3 to 2, 4 to 2, 7 to 2, 9 to 1)
+        val actualMap = FrequencyMapUsingMergeMethod(list)
+
+        assertEquals(expectedMap, actualMap)
+    }
+}
+fun frequencyMapUsingMutableMap(list: List<Int>): MutableMap<Int, Int> {
+    val map = mutableMapOf<Int, Int>()
+    for (value in list) {
+        val count = map.getOrDefault(value, 0)
+        map[value] = count + 1
+    }
+    return map
+}
+fun frequencyMapUsingTreeMapMethod(list: List<Int>): MutableMap<Int, Int>{
+    val map = TreeMap<Int, Int>()
+
+    list.forEach { element ->
+        map[element] = map.getOrDefault(element, 0) + 1
+    }
+
+    return map
+}
+
+fun frequencyMapUsingBinarySearchMethod(list: List<Int>): MutableMap<Int, Int>{
+    val sortedList = list.sorted()
+
+    val map = mutableMapOf<Int, Int>()
+
+    sortedList.forEach { element ->
```

Comment:
One small optimization here:

```suggestion
    sortedList.distinct().forEach { element ->
```

Reply:
Thanks for pointing that out.
Fixed!